modules/system/classes/extensions/ModuleManager.php

```diff
@@ -0,0 +1,283 @@
+                } elseif (false /* Detect if market */) {
+                    Preserver::instance()->store($extension);
+                    // @TODO: Update files from market
+                }
+            }
+        }
+
+        foreach ($modules as $module) {
+            $this->output->info(sprintf('Migrating %s module...', $module));
+            $this->migrator->run(base_path() . '/modules/' . strtolower($module) . '/database/migrations');
+
+            if ($firstUp) {
+                $className = '\\' . $module . '\Database\Seeds\DatabaseSeeder';
+                if (class_exists($className)) {
+                    $this->output->info(sprintf('Seeding %s module...', $module));
+
+                    $seeder = App::make($className);
+                    $return = $seeder->run();
+
+                    if ($return && (is_string($return) || is_array($return))) {
+                        $return = is_string($return) ? [$return] : $return;
+                        foreach ($return as $item) {
+                            $this->output->info(sprintf('[%s]: %s', $className, $item));
+                        }
+                    }
+
+                    $this->output->info(sprintf('Seeded %s', $module));
+                }
+            }
+        }
+
+        Parameter::set('system::update.count', 0);
+        CacheHelper::clear();
+
+        return true;
+    }
+
+    public function refresh(WinterExtension|string|null $extension = null): mixed
+    {
+        $this->rollback($extension);
+        return $this->update($extension, migrationsOnly: true);
+    }
+
+    /**
+     * @throws ApplicationException
+     */
+    public function rollback(WinterExtension|string|null $extension = null, ?string $targetVersion = null): mixed
+    {
+        $modules = $this->getModuleList($extension);
+
+        $paths = [];
+        foreach ($modules as $module) {
+            $paths[] = base_path() . '/modules/' . strtolower($module) . '/database/migrations';
+        }
+
+        while (true) {
+            $rolledBack = $this->migrator->rollback($paths, ['pretend' => false]);
+
+            if (count($rolledBack) == 0) {
+                break;
+            }
+        }
+
+        return true;
+    }
+
+    public function uninstall(WinterExtension|string $extension, bool $noRollback = false, bool $preserveFiles = false): mixed
+    {
+        if (!($module = $this->resolve($extension))) {
+            throw new ApplicationException(sprintf(
+                'Unable to uninstall module: %s',
+                is_string($extension) ? $extension : $extension->getIdentifier()
+            ));
+        }
+
+        if (!$noRollback) {
+            $this->rollback($module);
+        }
+
+        if (!$preserveFiles) {
+            // Modules probably should not be removed
+            // File::deleteDirectory($module->getPath());
```

Comment:
When would this method be called and the files not need to be removed?

Reply:
Also something to think about with the uninstall() method is removal from the composer.json

Reply:
When you run `./artisan winter:down && ./artisan winter:up` and he won't come up because you just deleted system :'(

Reply:
Right, but shouldn't those commands be passing the preserve files flag?